runtime/commonMain/src/kotlinx/serialization/json/internal/WriteMode.kt

```diff
@@ -15,15 +21,26 @@ internal enum class WriteMode(@JvmField val begin: Char, @JvmField val end: Char
     val endTc: Byte = charToTokenClass(end)
 }
 
-internal fun switchMode(desc: SerialDescriptor, typeParams: Array<out KSerializer<*>>): WriteMode =
+internal fun Json.switchMode(desc: SerialDescriptor): WriteMode =
     when (desc.kind) {
         UnionKind.POLYMORPHIC -> WriteMode.POLY_OBJ
         StructureKind.LIST -> WriteMode.LIST
-        StructureKind.MAP -> {
-            val keyKind = typeParams[0].descriptor.kind
-            if (keyKind is PrimitiveKind || keyKind == UnionKind.ENUM_KIND)
-                WriteMode.MAP
-            else WriteMode.LIST
-        }
+        StructureKind.MAP -> selectMapMode(desc, { WriteMode.MAP }, { WriteMode.LIST })
         else -> WriteMode.OBJ
     }
+
+internal inline fun <T, R1 : T, R2 : T> Json.selectMapMode(
+    mapDescriptor: SerialDescriptor,
+    ifMap: () -> R1,
+    ifList: () -> R2
+): T {
+    val keyDescriptor = mapDescriptor.getElementDescriptor(0)
+    val keyKind = keyDescriptor.kind
+    return if (keyKind is PrimitiveKind || keyKind == UnionKind.ENUM_KIND) {
```

Comment:
Should this place be tweaked when we support object serialization?

Reply:
Probably, I'll take a look at it